### PR TITLE
Update the Tactic interface

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Updated the `Tactic` interface.
* Separated the `updateParams` functionality into its own function
* The `Robot` is now assigned/updated through its own function rather than being given with the Parameters
* How `done()` is calculated has been slightly changed to detect when the `getNextIntent()` function returns a `nullptr`. This is to make sure that Tactics can't return a `nullptr` without reporting `done()`
* Updated tests and separated generic `Tactic` tests into their own file

### Testing Done

* Removed a few redundant unit tests
* Updated tests to use the new interface
* Added a new test to make sure `Tactics` and `Actions` don't return a `nullptr` without `done()` returning `true`

### Resolved Issues

None, but required for #394

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/tactic/move_tactic.cpp [hunk 1]
```diff
@@ -2,38 +2,33 @@
...
-    this->robot             = robot;
     this->destination       = destination;
     this->final_orientation = final_orientation;
     this->final_speed       = final_speed;
-
-    return getNextIntent();
 }
 
-double MoveTactic::calculateRobotCost(const Robot &robot, const Field &field)
+double MoveTactic::calculateRobotCost(const Robot &robot, const World &world)
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/move_tactic.cpp:16]: Why this change?

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h [hunk 2]
```diff
@@ -18,19 +19,35 @@ class Tactic
 {
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.h:33]: Could we give a high level overview in the javadoc function for this class of how we intend it to be used? (most concrete implementations will have an `updateParams` function, etc.)

File: src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h [hunk 3]
```diff
@@ -6,34 +6,44 @@
 class MoveTactic : public Tactic
 {
    public:
-    explicit MoveTactic(const Robot& robot);
+    /**
+     * Creates a new MoveTactic
+     *
+     * The MoveTactic will move the assigned robot to the given destination and arrive
+     * with the specified final orientation and speed
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h]: nit; but I think this description of what the `MoveTactic` does should be in the class Javadoc comment, rather then the constructor

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp [hunk 4]
```diff
@@ -11,18 +13,46 @@ bool Tactic::done() const
...
+    //
+    // We also check out internal done_ variable. This is set automatically by the
+    // getNextIntent function if we ever return a nullptr. We do this extra check because
+    // Tactics can return a nullptr one "tick" (aka function call) before the coroutine
+    // actually evaluates as done. This way we ensure that the Tactic::done() function
+    // will always evaluate to true on the exact same call a nullptr is returned.
+    //
+    // This is important since higher-level functions rely on the done() function, and
+    // returning nullptr values "out of sync" with this could cause problems
+    return done_ || !static_cast<bool>(intent_sequence);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp]: Why do we check `!static_cast<bool>(intent_sequence)` here if `done_` is already set?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, makes sense.

File: src/thunderbots/software/ai/hl/stp/tactic/move_tactic.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/move_tactic.cpp:16]: Future-proofing the code. I expect any goalie tactics to need to know about which robot is assigned to be the goalie according to the refbox, and since we're already using the `Team` I figured I'd just make it the `World`. This function will be called from the `PlayExecutor` level anyway which will already have a `World`.

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp]: The coroutine could complete without a nullptr being returned first, in which case the `done_` variable wouldn't be set.

Alternatively we can also set the `done_` variable after the `if(couroutine)` check anyway (where we default to returning a nullptr at the end of the function.

```
getNextIntent() {
    if(coroutine) {
        // Set done_ to true if a nullptr was returned
        // return result
    }
    done_ = true;
    return nullptr;
}
```

Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp]: Interesting, what's the difference between returning the nullptr and just completing? What would be the case for both? 

Either way, I think what you just suggested (setting `done_` in all cases) is a good option!

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h:33]: comment added

File: src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h]: comment moved

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp]: There's not much of a difference between them (both indicate we are "done"), but implementation details can cause a nullptr to be returned one iteration/tick before the `done()` function evaluates to `true`. We want these to always be perfectly in sync.

```
MoveAction move_action = MoveAction(*robot);
do
{
    yield(move_action.updateStateAndGetNextIntent(*robot, destination, final_orientation, final_speed));
} while (!move_action.done());
```

In this example, a nullptr will be returned one iteration before the coroutine is done. This is because once the action is done, one more `Action::getNextIntent` will be yielded before we see the action is done and finish. This is the case with `while` loops as well, because `Action::getNextIntent` has to be called a second time for the Action to realize it's done, so we will always return a nullptr one iteration before the coroutine finishes.

On the other hand, if the Tactic was directly constructing and yielding Intents (which is valid), then the coroutine would finish normally without prematurely returning any nullptrs.

Hope that makes things clear. Regardless, I've made the change I suggested above (setting the `done_` variable again and just using that)

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h [hunk 5]
```diff
@@ -13,24 +14,54 @@
  * and they tend to rely a lot on our Evaluation functions. Ultimately, Tactics will
  * return the next Intent that the Robot assigned to this Tactic should run in order
  * to work towards its objective.
+ *
+ * HOW THIS CLASS IS USED:
+ * Plays will construct and return the Tactics they want to be running. Every time a play
+ * is run, it will update the parameters of each tactic with the updateParams(...)
+ * function (see the concrete implementations of this class for examples). This is done
+ * every time in order fo the Tactics to have the most up to date information when they
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.h:22]:
``` suggestion
 * every time in order for the Tactics to have the most up to date information when they
```

**Review by garethellis0 — APPROVED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 5 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h:22]: Fixing this in my next PR
